protocol/options.go

```diff
@@ -0,0 +1,73 @@
+package protocol
+
+import (
+	"fmt"
+
+	"github.com/libp2p/go-smart-record/ir"
+	"github.com/libp2p/go-smart-record/ir/base"
+)
+
+// ModeOpt publicly exposes the modes in which the smartRecordManager can operate
+type ModeOpt int
+
+const (
+	ModeClient ModeOpt = iota
+	ModeServer
+)
+
+// Options is a structure containing all the options that can be used when constructing the smart records env
+type config struct {
+	//datastore          ds.Batching
+	updateContext ir.UpdateContext
+	assembler     ir.Assembler
+	mode          ModeOpt
+}
+
+// Option type for smart records
+type Option func(*config) error
+
+// defaults are the default smart record env options. This option will be automatically
+// prepended to any options you pass to the constructor.
+var defaults = func(o *config) error {
+	o.updateContext = ir.DefaultUpdateContext{}
+	o.assembler = base.BaseGrammar
+	o.mode = ModeServer
+
+	return nil
+}
+
+// apply applies the given options to this Option
+func (c *config) apply(opts ...Option) error {
+	for i, opt := range opts {
+		if err := opt(c); err != nil {
+			return fmt.Errorf("smart record option %d failed: %s", i, err)
+		}
+	}
+	return nil
+}
+
+// Mode configures which mode the smartRecordManager operates in (Client, Server).
+//
+// Defaults to Server.
+func Mode(m ModeOpt) Option {
```

Comment:
We misunderstood each other when we said that the "client and server functionality should be separated":
- You've separated them at runtime: A mode option tells the same type, SmartRecordManager, whether to behave as a client or a server.
- I meant they need to be separate at compile time. Put differently, there should be two independent types, one SmartRecordClient and one SmartRecordServer.

OK, so why is that? Many reasons. For instance:

- There can/will be multiple implementations of a server (perhaps optimized for different use cases), but the client library implementation wouldn't change (because the server and the client are decoupled by the protocol between them, which doesn't change). So, there is no reason to tie the client code with any one particular implementation of the server, which is what you have right now.

- Another way of saying this: You should be able to distribute the client library independently of the server implementations. And the server implementations should be distributable independently of the client library.

Reply:
Oh, I see!! Makes total sense. I thought about that but did it in this way so I didn't have to change much in the libp2p and go-ipfs integrations. No problem I will change this tomorrow and ping you when is ready.